### PR TITLE
Track API requests

### PR DESCRIPTION
**Pull request by mnacos**
### Context

As part of the alpha API release, but also post-launch, we'll need to be able to track API usage for various reasons: technical (performance & capacity planning) but also operational (onboarding & engagement tracking).

During the initial requirement capture, we have split requirements into technical and business:

Technical requirements:

- fast writes - no significant slowdown of API performance (numbers?)
- query interface for analysis (how realtime? logging format?)
- storage: app logs vs. db table vs. external hosted log service (e.g. logit) vs. adding to our stack (e.g. redis)
- two types of tracking: usage (not there) vs. error (we may have this already)
- reporting: what should be automatic and what can we do manually

Business requirements (insights required):

- are vendors using the API? [* - requests/period/vendor]
  - which endpoints?
- are they using it correctly? [* - most common errors by frequency]
- how much are they using it? [*]
  - relative to one another [*]
- tokens per vendor
  - users per token
- are providers using the API?
- time from token generation to first request [* - manual report]

Starred items are the only ones relevant for the alpha API release.

### Changes proposed in this pull request

Default development logs will stay the same to assist development, but production logs will switch to lograge (one line per request) and the logstash json format, so that they can be easily indexed and queried within an ELK stack (we're targetting Logit).

To satisfy the inital goal of this story, which is to track API usage per api key and provider, additional metadata will be added to the lograge/logstash format, in particular: vendor_api_token_id, provider_id.

Considering this is app-wide, I have also added a candidate_id, while suppressing candidate params, to avoid accidental capture of PII data.

DOMAIN needs to be set, as we'll probably use the same Logit stack to receive logs from multiple environments (production, qa, sandbox)

Production logs will be sent to both STDOUT, which is considered good practice for Docker-based deployments (allowing docker to send logs to a centralised logging service) and a Logstash/Logit server via TCP. The logstash-logger gem, responsible for TCP shipping, seems to recover from transient networking errors (it uses a background thread), but the possibility for some log data loss is there. We have agreed this is acceptable risk for now, as we are planning to use Logit to track API usage and assist the onboarding of vendors/providers.

### Guidance to review

In development mode, you can try out the lograge/logstash format used for production by setting the following environment variable in .env.development and restarting Rails:

```
LOGSTASH_ENABLE=true
```

You can also try out shipping logs to Logit by setting LOGSTASH_REMOTE to 'true'. Please add the following to your env:

```
LOGSTASH_ENABLE=true
LOGSTASH_REMOTE=true
LOGSTASH_HOST=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx-ls.logit.io
LOGSTASH_PORT=xxxxx
LOGSTASH_SSL=true
```

where 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx-ls.logit.io' is a url specific to your logit stack. It seems that the id of the stack plus the port number (which changes across stacks) is what provides some security for Logit when using TCP for submitting logs.

You can then interact with your local app and observe what is sent to Logit by browsing their Kibana interface. Please ask me or Duncan to add you to the relevant Logit account.

image

image

Things to try:

#### successful API request with API key

```$ curl-H 'Authorization: Bearer your_token' http://localhost:3000/api/v1/applications?since=2019-10-21T18:02:47+01:00```

image

Note vendor_api_token_id, provider and domain are set.

#### authenticated API request triggering 500 Internal Server Error

```$ curl-H 'Authorization: Bearer your_token' http://localhost:3000/api/v1/applications?since=derp```

image

Note some exception info, i.e. 'ArgumentError: invalid date' is included.

#### candidate sign-up

Please use the web UI for this.

image

Note that {candidate: {email_address: 'user@example.com'}} is suppressed.

### Link to Trello card

[1125 - Track API requests](https://trello.com/c/w2i50NdJ)

### CODE REVIEW

File: lib/logging.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+class Logging
```
Comment (duncanjbrown) [inline lib/logging.rb]: I like the way we've extracted this! It's not clear that it's not a built-in Rails class. Could we put a namespace on it, perhaps?

More minor: this is acting more like a module than a class — could it be one?

File: config/environments/production.rb [hunk 2]
```diff
@@ -72,17 +65,17 @@
...
-
-  # Use a different logger for distributed setups.
-  # require 'syslog/logger'
-  # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new 'app-name')
-
-  if ENV["RAILS_LOG_TO_STDOUT"].present?
-    logger           = ActiveSupport::Logger.new(STDOUT)
+  # Logging configuration
+  if ENV['LOGSTASH_ENABLE'] == 'true'
+    require './lib/logging'
```
Comment (duncanjbrown) [inline config/environments/production.rb]: I think autoload might have our back here, but not sure?

File: config/environments/production.rb [hunk 3]
```diff
@@ -72,17 +65,17 @@
...
-  config.log_formatter = ::Logger::Formatter.new
-
-  # Use a different logger for distributed setups.
-  # require 'syslog/logger'
-  # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new 'app-name')
-
-  if ENV["RAILS_LOG_TO_STDOUT"].present?
-    logger           = ActiveSupport::Logger.new(STDOUT)
+  # Logging configuration
+  if ENV['LOGSTASH_ENABLE'] == 'true'
```
Comment (duncanjbrown) [inline config/environments/production.rb]: Given that we look up an env var called `LOGSTASH_ENABLE` here, does it make sense for the implementation we call to have a generic name like `Logging`?

File: lib/logging.rb (hunk 1 above)
Comment (mnacos) [reply lib/logging.rb]: I have changed the name from Logging to LogstashLogging.

File: config/environments/production.rb (hunk 2 above)
Comment (mnacos) [reply config/environments/production.rb]: My local environment dies without the explicit require there.

File: config/environments/production.rb (hunk 3 above)
Comment (mnacos) [reply config/environments/production.rb]: Good point, I've changed it to the slightly more specific LogstashLogging.

File: Gemfile [hunk 4]
```diff
@@ -37,6 +37,11 @@ gem 'actionview-component'
 
 gem 'uk_postcode'
 
+# Logging
```
Comment (duncanjbrown) [inline Gemfile]: Nice

File: app/controllers/candidate_interface/candidate_interface_controller.rb [hunk 5]
```diff
@@ -4,7 +4,14 @@ class CandidateInterfaceController < ActionController::Base
     before_action :require_basic_auth_for_ui
     before_action :authenticate_candidate!
     layout 'application'
-
     alias :audit_user :current_candidate
+
+  private
+
+    # contoller-specific additional info to included in lograge/logstash logs
```
Comment (duncanjbrown) [inline app/controllers/candidate_interface/candidate_interface_controller.rb]: Typo (here and elsewhere): `contoller`

File: app/lib/azure_environment.rb [hunk 6]
```diff
@@ -4,6 +4,8 @@ def self.authorised_hosts
   end
 
   def self.hostname
-    ENV.fetch('CUSTOM_HOSTNAME', authorised_hosts.first)
+    hostname = authorised_hosts.first rescue nil
+    hostname ||= Socket.gethostname
+    ENV.fetch('CUSTOM_HOST_NAME', hostname)
```
Comment (duncanjbrown) [inline app/lib/azure_environment.rb]: This will need a rebase (another!) against #408 where we've now renamed the env var

Comment (duncanjbrown) [reply app/lib/azure_environment.rb]: I like falling back to `Socket.hostname` but I feel like this code might not belong inside the `AzureEnvironment` module. I *think* I'd rather have 

```
domain: Rails.env.production? ? AzureEnvironment.hostname : Socket.hostname
```

in `LogStashLogging`? This module is then 100% about Azure and the dependency on those ENV vars is clear. What do you think?

Comment (mnacos) [reply app/lib/azure_environment.rb]: good call, updating

**Pull request merged**